File: src/GpuData/Core+PCache1/10.PCache1.cpp
namespace Core
{
	struct PCache1
	{
		// Cache configuration parameters. Page size (szPage) and the purgeable flag (bPurgeable) are set when the cache is created. nMax may be 
		// modified at any time by a call to the pcache1Cachesize() method. The PGroup mutex must be held when accessing nMax.
		PGroup *Group;					// PGroup this cache belongs to
		int SizePage;                   // Size of allocated pages in bytes
		int SizeExtra;                  // Size of extra space in bytes
		bool Purgeable;					// True if cache is purgeable
		unsigned int Min;				// Minimum number of pages reserved
		unsigned int Max;				// Configured "cache_size" value
		unsigned int N90pct;			// nMax*9/10
		unsigned int MaxKey;			// Largest key seen since xTruncate()

		// Hash table of all pages. The following variables may only be accessed when the accessor is holding the PGroup mutex.
		unsigned int Recyclables;       // Number of pages in the LRU list
		unsigned int Pages;             // Total number of pages in apHash
		unsigned int _0; PgHdr1 **Hash;	// Hash table for fast lookup by key
	};

#define PCache1_EnterMutex(x) MutexEx_Enter((x)->Mutex)
#define PCache1_LeaveMutex(x) MutexEx_Leave((x)->Mutex)
}

File: src/GpuData/Core+PCache1/10.PCache1.cs
using Pid = System.UInt32;
namespace Core
{
    public partial class PCache1
    {
        // Cache configuration parameters. Page size (szPage) and the purgeable flag (bPurgeable) are set when the cache is created. nMax may be 
        // modified at any time by a call to the pcache1CacheSize() method. The PGroup mutex must be held when accessing nMax.
        public PGroup Group;        // PGroup this cache belongs to
        public int SizePage;        // Size of allocated pages in bytes
        public int SizeExtra;       // Size of extra space in bytes
        public bool Purgeable;      // True if cache is purgeable
        public uint Min;             // Minimum number of pages reserved
        public uint Max;             // Configured "cache_size" value
        public uint N90pct;          // nMax*9/10
        public Pid MaxKey;          // Largest key seen since xTruncate()
        
        // Hash table of all pages. The following variables may only be accessed when the accessor is holding the PGroup mutex.
        public uint Recyclables;    // Number of pages in the LRU list
        public uint Pages;          // Total number of pages in apHash
        public PgHdr1[] Hash;       // Hash table for fast lookup by key

        public void Clear()
        {
            Recyclables = 0;
            Pages = 0;
            Hash = null;
            MaxKey = 0;
        }

        //private static PgHdr PGHDR1_TO_PAGE(PgHdr1 p) { return p.Page; }
        //private static PgHdr1 PAGE_TO_PGHDR1(PCache1 c, PgHdr p) { return (PgHdr1)p.PgHdr1; }
        private static void PCache1_EnterMutex(PGroup x) { MutexEx.Enter(x.Mutex); }
        private static void PCache1_LeaveMutex(PGroup x) { MutexEx.Leave(x.Mutex); }
    }
}
